producer.go

```diff
@@ -0,0 +1,343 @@
+// variables from the config parameter, or returning an non-nil error if
+// some of the configuration variables were invalid.
+func StartProducer(config ProducerConfig) (p *Producer, err error) {
+	if len(config.Topic) == 0 {
+		err = errors.New("creating a producer requires a non-empty topic")
+		return
+	}
+
+	if len(config.Address) == 0 {
+		config.Address = "localhost:4151"
+	}
+
+	if config.MaxConcurrency == 0 {
+		config.MaxConcurrency = DefaultMaxConcurrency
+	}
+
+	if config.DialTimeout == 0 {
+		config.DialTimeout = DefaultDialTimeout
+	}
+
+	if config.ReadTimeout == 0 {
+		config.ReadTimeout = DefaultReadTimeout
+	}
+
+	if config.WriteTimeout == 0 {
+		config.WriteTimeout = DefaultWriteTimeout
+	}
+
+	if config.MaxRetryTimeout == 0 {
+		config.MaxRetryTimeout = DefaultMaxRetryTimeout
+	}
+
+	if config.MinRetryTimeout == 0 {
+		config.MinRetryTimeout = DefaultMinRetryTimeout
+	}
+
+	p = &Producer{
+		reqs:            make(chan ProducerRequest, config.MaxConcurrency),
+		done:            make(chan struct{}),
+		address:         config.Address,
+		topic:           config.Topic,
+		dialTimeout:     config.DialTimeout,
+		readTimeout:     config.ReadTimeout,
+		writeTimeout:    config.WriteTimeout,
+		maxRetryTimeout: config.MaxRetryTimeout,
+		minRetryTimeout: config.MinRetryTimeout,
+	}
+	p.join.Add(config.MaxConcurrency)
+
+	for i := 0; i != config.MaxConcurrency; i++ {
+		go p.run()
+	}
+
+	return
+}
+
+// Stop gracefully shutsdown the producer, cancelling all inflight requests and
+// waiting for all backend connections to be closed.
+//
+// It is safe to call the method multiple times and from multiple goroutines,
+// they will all block until the producer has been completely shutdown.
+func (p *Producer) Stop() {
+	p.once.Do(p.stop)
+	err := errors.New("publishing to a producer that was already stopped")
+
+	for req := range p.reqs {
+		req.complete(err)
+	}
+
+	p.join.Wait()
+}
+
+// Publish sends a message using the producer p, returning an error if it was
+// already closed or if an error occurred while publishing the message.
+//
+// Note that no retry is done internally, the producer will fail after the
+// first unsuccessful attempt to publish the message. It is the responsibility
+// of the caller to retry if necessary.
+func (p *Producer) Publish(message []byte) (err error) {
+	defer func() {
+		if recover() != nil {
+			err = errors.New("publishing to a producer that was already stopped")
+		}
+	}()
+
+	res := make(chan error, 1)
+
+	p.reqs <- ProducerRequest{
+		Message:  message,
+		Response: res,
+	}
+
+	err = <-res
+	return
+}
+
+// Requests returns a write-only channel that can be used to submit requests to p.
+//
+// This method is useful when the publish operation needs to be associated with
+// other operations on channels in a select statement for example, or to publish
+// in a non-blocking fashion.
+func (p *Producer) Requests() chan<- ProducerRequest {
```

Comment:
I kind of agree but at the same time I'm gonna feel annoyed when I'll have to write 50 lines instead of 1 to get around the fact that we didn't expose that method. Like the documentation says this is not the preferred way of using the producer but it is a convenience interface when the code is designed around channels.

I'm not 100% sure it's the right choice but I know I'm often frustrated when this kind of API isn't available.

Feel free to discuss if you feel strongly about it.